e2e/e2e_test.go

```diff
@@ -0,0 +1,2610 @@
+	}
+	if err := newCmd(t, "deploy").Run(); err != nil {
+		t.Fatal(err)
+	}
+	if !waitForContent(t, fmt.Sprintf("http://%v.default.localtest.me", name), "UPDATED") {
+		t.Fatalf("function did not update correctly")
+	}
+}
+
+// TestCore_Describe ensures that describing a function accurately represents
+// its expected state.
+//
+//	func describe
+func TestCore_Describe(t *testing.T) {
+	name := "func-e2e-test-core-describe"
+	_ = fromCleanEnv(t, name)
+
+	if err := newCmd(t, "init", "-l=go").Run(); err != nil {
+		t.Fatal(err)
+	}
+
+	cmd := newCmd(t, "deploy")
+	if err := cmd.Start(); err != nil {
+		t.Fatal(err)
+	}
+	defer func() {
+		clean(t, name, DefaultNamespace)
+	}()
+
+	if err := cmd.Wait(); err != nil {
+		t.Fatalf("deploy error. %v", err)
+	}
+
+	if !waitForEcho(t, fmt.Sprintf("http://%v.default.localtest.me", name)) {
+		t.Fatalf("function did not deploy correctly")
+	}
+
+	// Call func describe with JSON output
+	cmd = newCmd(t, "describe", "--output=json")
+	out := bytes.Buffer{}
+	cmd.Stdout = &out
+	if err := cmd.Run(); err != nil {
+		t.Fatal(err)
+	}
+
+	// Parse the JSON output
+	var instance fn.Instance
+	if err := json.Unmarshal(out.Bytes(), &instance); err != nil {
+		t.Fatalf("error unmarshaling describe output: %v", err)
+	}
+
+	// Validate that the name matches what we expect
+	if instance.Name != name {
+		t.Errorf("Expected name %q, got %q", name, instance.Name)
+	}
+}
+
+// TestCore_Invoke ensures that the invoke helper functions for both
+// local and remote function instances.
+//
+//	func invoke
+func TestCore_Invoke(t *testing.T) {
+	name := "func-e2e-test-core-invoke"
+	_ = fromCleanEnv(t, name)
+
+	if err := newCmd(t, "init", "-l=go").Run(); err != nil {
+		t.Fatal(err)
+	}
+
+	// Test local invocation
+	// ----------------------------------------
+	// Runs the function locally, which `func invoke` will invoke when
+	// it detects it is running.
+	address, err := chooseOpenAddress(t)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	cmd := newCmd(t, "run", "--address", address)
+	if err := cmd.Start(); err != nil {
+		t.Fatal(err)
+	}
+	run := cmd // for the closure
+	defer func() {
+		// ^C the running function
+		if err := run.Process.Signal(os.Interrupt); err != nil {
+			fmt.Fprintf(os.Stderr, "error interrupting. %v", err)
+		}
+	}()
+
+	// TODO: complete implementation of `func run --json` structured output
+	// such that we can parse it for the actual listen address in the case
+	// that there is already something else running on 8080
+	if !waitForEcho(t, "http://"+address) {
+		t.Fatalf("service does not appear to have started correctly.")
+	}
+
+	// Check invoke
+	cmd = newCmd(t, "invoke", "--data=func-e2e-test-core-invoke-local")
+	out := bytes.Buffer{}
+	cmd.Stdout = &out
+	if err := cmd.Run(); err != nil {
+		t.Fatal(err)
+	}
+	if !strings.Contains(out.String(), "func-e2e-test-core-invoke-local") {
+		t.Logf("out: %v", out.String())
+		t.Fatal("function invocation did not echo data provided")
+	}
```

Comment:
Put in a function and use here and below to avoid duplication.

```suggestion
	checkInvoke := func(funcName string) {
		cmd = newCmd(t, "invoke", "--data="+funcName)
		out := bytes.Buffer{}
		cmd.Stdout = &out
		if err := cmd.Run(); err != nil {
			t.Fatal(err)
		}
		if !strings.Contains(out.String(), funcName) {
			t.Logf("out: %v", out.String())
			t.Fatal("function invocation did not echo data provided")
		}
	}
	checkInvoke("func-e2e-test-core-invoke-local")
```